File: desktop/tests/subagent-index.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { SubagentIndex } from '../src/main/subagent-index';

describe('SubagentIndex', () => {
  let idx: SubagentIndex;

  beforeEach(() => {
    idx = new SubagentIndex({ nowMs: () => 1000 });
  });

  it('binds a subagent to the most recent matching parent', () => {
    idx.recordParentAgentToolUse('toolu_A', 'Find bug', 'Explore');
    const bound = idx.bindSubagent('agent1', { description: 'Find bug', agentType: 'Explore' });
    expect(bound).toBe('toolu_A');
    expect(idx.lookup('agent1')).toBe('toolu_A');
  });

  it('returns null when no parent matches', () => {
    const bound = idx.bindSubagent('agent1', { description: 'Find bug', agentType: 'Explore' });
    expect(bound).toBeNull();
    expect(idx.lookup('agent1')).toBeNull();
  });

  it('FIFO: two parallel parents with identical description bind in emit order', () => {
    idx.recordParentAgentToolUse('toolu_A', 'Review diff', 'general-purpose');
    idx.recordParentAgentToolUse('toolu_B', 'Review diff', 'general-purpose');

    const bound1 = idx.bindSubagent('agent1', { description: 'Review diff', agentType: 'general-purpose' });
    const bound2 = idx.bindSubagent('agent2', { description: 'Review diff', agentType: 'general-purpose' });

    expect(bound1).toBe('toolu_A');
    expect(bound2).toBe('toolu_B');
  });

  it('does not match parents with different subagent_type', () => {
    idx.recordParentAgentToolUse('toolu_A', 'Do stuff', 'Explore');
    const bound = idx.bindSubagent('agent1', { description: 'Do stuff', agentType: 'Plan' });
    expect(bound).toBeNull();
  });

  it('binding consumes the parent so it is not reused', () => {
    idx.recordParentAgentToolUse('toolu_A', 'Find bug', 'Explore');
    idx.bindSubagent('agent1', { description: 'Find bug', agentType: 'Explore' });
    const second = idx.bindSubagent('agent2', { description: 'Find bug', agentType: 'Explore' });
    expect(second).toBeNull();
  });

  it('unbind removes a binding so lookup returns null', () => {
    idx.recordParentAgentToolUse('toolu_A', 'Find bug', 'Explore');
    idx.bindSubagent('agent1', { description: 'Find bug', agentType: 'Explore' });
    idx.unbind('agent1');
    expect(idx.lookup('agent1')).toBeNull();
  });

  it('pending subagent events buffer then flush when parent arrives later', () => {
    idx.bufferPendingEvent('agent1', { description: 'Find bug', agentType: 'Explore' }, { fakeEvent: 1 });
    idx.bufferPendingEvent('agent1', { description: 'Find bug', agentType: 'Explore' }, { fakeEvent: 2 });

    idx.recordParentAgentToolUse('toolu_A', 'Find bug', 'Explore');

    const flushed = idx.tryFlushPending('agent1');
    expect(flushed?.parentToolUseId).toBe('toolu_A');
    expect(flushed?.events).toEqual([{ fakeEvent: 1 }, { fakeEvent: 2 }]);
    expect(idx.lookup('agent1')).toBe('toolu_A');
  });

  it('pending events age out after 30s with no matching parent', () => {
    const clock = { t: 1000 };
    const aged = new SubagentIndex({ nowMs: () => clock.t });

    aged.bufferPendingEvent('agent1', { description: 'Find bug', agentType: 'Explore' }, { fakeEvent: 1 });
    clock.t = 1000 + 30_001;
    aged.pruneExpired();

    expect(aged.tryFlushPending('agent1')).toBeNull();
  });

  it('pruneExpired keeps entries younger than 30s', () => {
    const clock = { t: 1000 };
    const aged = new SubagentIndex({ nowMs: () => clock.t });
    aged.bufferPendingEvent('agent1', { description: 'Find bug', agentType: 'Explore' }, { fakeEvent: 1 });
    clock.t = 1000 + 15_000;
    aged.pruneExpired();

    aged.recordParentAgentToolUse('toolu_A', 'Find bug', 'Explore');
    expect(aged.tryFlushPending('agent1')?.parentToolUseId).toBe('toolu_A');
  });
});

File: app/src/test/kotlin/com/youcoded/app/parser/SubagentIndexTest.kt
package com.youcoded.app.parser

import org.junit.Assert.*
import org.junit.Test

class SubagentIndexTest {

    @Test
    fun `binds subagent to matching parent`() {
        val idx = SubagentIndex(nowMs = { 1000L })
        idx.recordParentAgentToolUse("toolu_A", "Find bug", "Explore")
        val bound = idx.bindSubagent("agent1", description = "Find bug", agentType = "Explore")
        assertEquals("toolu_A", bound)
        assertEquals("toolu_A", idx.lookup("agent1"))
    }

    @Test
    fun `returns null when no parent matches`() {
        val idx = SubagentIndex()
        val bound = idx.bindSubagent("agent1", description = "Find bug", agentType = "Explore")
        assertNull(bound)
    }

    @Test
    fun `FIFO pairing for parallel parents with identical description`() {
        val idx = SubagentIndex()
        idx.recordParentAgentToolUse("toolu_A", "Review", "general-purpose")
        idx.recordParentAgentToolUse("toolu_B", "Review", "general-purpose")
        assertEquals("toolu_A", idx.bindSubagent("a1", "Review", "general-purpose"))
        assertEquals("toolu_B", idx.bindSubagent("a2", "Review", "general-purpose"))
    }

    @Test
    fun `subagent_type mismatch means no binding`() {
        val idx = SubagentIndex()
        idx.recordParentAgentToolUse("toolu_A", "Do stuff", "Explore")
        assertNull(idx.bindSubagent("agent1", "Do stuff", "Plan"))
    }

    @Test
    fun `binding consumes the parent so it is not reused`() {
        val idx = SubagentIndex()
        idx.recordParentAgentToolUse("toolu_A", "Find bug", "Explore")
        idx.bindSubagent("agent1", "Find bug", "Explore")
        val second = idx.bindSubagent("agent2", "Find bug", "Explore")
        assertNull(second)
    }

    @Test
    fun `unbind clears binding`() {
        val idx = SubagentIndex()
        idx.recordParentAgentToolUse("toolu_A", "Find bug", "Explore")
        idx.bindSubagent("agent1", "Find bug", "Explore")
        idx.unbind("agent1")
        assertNull(idx.lookup("agent1"))
    }

    @Test
    fun `pending events flush when parent arrives`() {
        val idx = SubagentIndex()
        idx.bufferPendingEvent("agent1", "Find bug", "Explore", "event1")
        idx.bufferPendingEvent("agent1", "Find bug", "Explore", "event2")
        idx.recordParentAgentToolUse("toolu_A", "Find bug", "Explore")
        val flushed = idx.tryFlushPending("agent1")
        assertNotNull(flushed)
        assertEquals("toolu_A", flushed!!.parentToolUseId)
        assertEquals(listOf<Any>("event1", "event2"), flushed.events)
    }

    @Test
    fun `pending events age out after 30s`() {
        var clock = 1000L
        val idx = SubagentIndex(nowMs = { clock })
        idx.bufferPendingEvent("agent1", "Find bug", "Explore", "event1")
        clock += 30_001L
        idx.pruneExpired()
        assertNull(idx.tryFlushPending("agent1"))
    }
}
